Mathlib/Logic/Godel/GodelBetaFunction.lean

```diff
@@ -0,0 +1,173 @@
+def listSup (l : List ℕ) := max l.length (List.foldr max 0 l) + 1
+
+/-- A list of coprimes -/
+def coprimeList (l : List ℕ) : List (ℕ × ℕ) :=
+  List.ofFn (fun i : Fin l.length => (l.get i, (i + 1) * (listSup l)! + 1))
+
+@[simp] lemma coprimeList_length (l : List ℕ) : (coprimeList l).length = l.length :=
+  by simp[coprimeList]
+
+lemma coprimeList_lt (l : List ℕ) (i) : ((coprimeList l).get i).1 < ((coprimeList l).get i).2 := by
+  have h₁ : l.get (i.cast $ by simp[coprimeList]) < listSup l :=
+    Nat.lt_add_one_iff.mpr (by simpa using Or.inr $ List.le_max_of_le (List.get_mem _ _ _) (by rfl))
+  have h₂ : Nat.listSup l ≤ (i + 1) * (Nat.listSup l)! + 1 :=
+    le_trans (self_le_factorial _) (le_trans (Nat.le_mul_of_pos_left (succ_pos _))
+      (le_add_right _ _))
+  simpa only [coprimeList, List.get_ofFn] using lt_of_lt_of_le h₁ h₂
+
+lemma coprime_mul_succ {n m a} (h : n ≤ m) (ha : m - n ∣ a) : Coprime (n * a + 1) (m * a + 1) :=
+  Nat.coprime_of_dvd (by
+    intro p pp hn hm
+    have : p ∣ (m - n) * a := by
+      simpa [Nat.succ_sub_succ, ← Nat.mul_sub_right_distrib] using
+        Nat.dvd_sub (Nat.succ_le_succ $ Nat.mul_le_mul_right a h) hm hn
+    have : p ∣ a := by
+      rcases (Nat.Prime.dvd_mul pp).mp this with (hp | hp)
+      · exact Nat.dvd_trans hp ha
+      · exact hp
+    have : p = 1 := by
+      simpa[Nat.add_sub_cancel_left] using Nat.dvd_sub (le_add_right _ _) hn
+        (Dvd.dvd.mul_left this n)
+    simp[this] at pp
+    apply not_prime_one at pp
+    exact pp)
+
+lemma coprimes_coprimeList (l : List ℕ) : Coprimes ((coprimeList l).map Prod.snd) := by
+  have : (coprimeList l).map Prod.snd =
+      List.ofFn (fun i : Fin l.length => (i + 1) * (listSup l)! + 1) := by
+    simp[coprimeList, Function.comp]
+  rw[this]
+  exact coprimes_of_nodup
+    (List.nodup_ofFn_ofInjective $ by
+       intro i j; simp[listSup, ← Fin.ext_iff, Nat.factorial_ne_zero])
+    (by
+      simp[← Fin.ext_iff, not_or]
+      suffices : ∀ i j : Fin l.length, i < j → Coprime ((i + 1) * (listSup l)! + 1) ((j + 1) *
+        (listSup l)! + 1)
+      · intro i j hij _
+        rcases Ne.lt_or_lt hij with (hij | hij)
+        · exact this i j hij
+        · exact coprime_comm.mp (this j i hij)
+      intro i j hij
+      have hjl : j < listSup l := lt_of_lt_of_le j.prop (le_step (le_max_left l.length
+        (List.foldr max 0 l)))
+      apply coprime_mul_succ
+        (Nat.succ_le_succ $ le_of_lt hij)
+        (Nat.dvd_factorial (by simp[Nat.succ_sub_succ, hij]) (by
+          simpa only [Nat.succ_sub_succ] using le_of_lt (lt_of_le_of_lt (sub_le j i) hjl))))
+
+/-- Gödel's Beta Function -/
+def beta (n i : ℕ) := n.unpair.1 % ((i + 1) * n.unpair.2 + 1)
+
+/-- Inverse of Gödel's Beta Function -/
+def unbeta (l : List ℕ) :=
+  (chineseRemainderList (coprimeList l) (coprimes_coprimeList l) : ℕ).pair (listSup l)!
+
+/-- **Gödel's Beta Function Lemma** -/
+lemma beta_function_lemma (l : List ℕ) (i : Fin l.length) :
```

Comment:
We already have [List.encodable](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Logic/Equiv/List.html#List.encodable). How is this different from `encode`/`decode` for lists?

Reply:
Or perhaps even more closely related: [Equiv.listNatEquivNat](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Logic/Equiv/List.html#Equiv.listNatEquivNat)

Reply:
Goedel's beta function is used to prove that arithmetically definable functions are equal to `Partrec'`. ([proof](https://github.com/iehality/lean4-logic/blob/7a12751321928ed48c565b6db906d1d8a303ed25/Logic/Vorspiel/Arith.lean#L492)) While `List.encodable` is computable, it is not immediately apparent whether it is arithmetically definable.

Reply:
But it is arithmetically definable. `encodeList` and `decodeList` are defined in terms of `Nat.pair` and `Nat.unpair`, and these are arithmetically definable (I guess it might be hardest to prove that for `Nat.sqrt`). 
I think it would be nicer to prove this reusing existing Mathlib definitions.

Reply:
The term arithmetically definable function, we mean here a class of functions whose graph can be described by a $\Sigma_1$ formula of arithmetic.  While it is straightforward to prove that this class is closed under function composition, but whether it is closed in primitive recursion is a proposition demanding a technical proof (we use beta function lemma to prove this). Since `encodeList` and `decodeList` are defined using recursion, they cannot be used to prove this.

Reply:
OK I have added that to the documentation. Thanks Floris.